---
title: "An Introduction to BSL"
author: 
- Shiori Yabe
- Hiroyoshi Iwata
- Jean-Luc Jannink
date: "2018-02-08"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{An Introduction to BSL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(BreedingSchemeLanguage)
knitr::opts_chunk$set(fig.dpi = 96)
```

BreedingSchemeLanguage is a package that simulates plant breeding using phenotypic selection, genomic selection, and selection based on pedigree method. In this package, users can simulate phenotyping in multiple environment (e.g., year and location) correlating each other, calculate the cost for breeding, and compare multiple breeding strategies.
This vignette shows an example that users can simulate using BreedingSchemeLanguage.

## Start using BreedingSchemeLanguage

At first, we load the BreedingSchemeLanguage in R terminal.

```{r, eval = FALSE}
library(BreedingSchemeLanguage)
```

## Define simulation settings

Plant species and whole simulation settings are specified with:

- When we hope to simulate a new founder population for breeding, we can create a new population.

```{r}
simEnv <- defineSpecies(nSim = 5)
```


- Here, we plan to conduct five simulation trials using default settings (i.e., plant species with seven chromosomes with 150cM length, effective population size of breeding population is 100, and trait controlled by only additive QTLs, etc.). If we hope to simulate new breeding schemes using pre-simulated breeding settings, we can load the pre-simulated settings.

```{r}
simEnv <- defineSpecies(loadData = "previousData")
```

- Only in this function, we have to store the whole settings into 'simEnv'.

Environmental variance and breeding cost are specified with:

- We assume three locations that are correlated each other. The correlation coefficients are fixed here and correlation between location 1 and 2, 1 and 3, and 2 and 3 are 0.6, 0.3 and 0.8. This relationship is entered in matrix format.
- The variance of year is not fixed.
- We have two kinds of plots one with error variance 4, the other with error variance 1. The former costs 2 units the latter 5 units.

```{r}
locCor <- matrix(c(1, 0.6, 0.3, 0.6, 1, 0.8, 0.3, 0.8, 1), 3)
errVars <- c(Early=4, Late=1)
plotCosts <- c(Early=2, Late=5)
defineVariances(locCorrelations=locCor, plotTypeErrVars=errVars)
defineCosts(phenoCost=plotCosts)
```

The initial breeding population is set as:

-The initial breeding population is set including above settings. The population size is 100 (default) here.

```{r}
initializePopulation()
```

## Simulate breeding schemes

Obtain the genotypic values of individuald in the initial breeding population.

- Phenotype in all locations over two years in the less expensive way.
- To calculate the genotypic values from phenotypic data, we use 'No information sharing' option, which means that individual effects are IID

```{r}
phenotype(locations=1:3, years=1:2, plotType="Early")
predictValue()
```

Phenotypic selection

- Conduct phenotypic selection and mating among the selected parents. Here, we select 40 individuals (default) and assume random mating. As the next generation, 100 individuals are generated (default).

```{r}
select()
cross()
```

Selection based on pedigree information

- We assume the selection held in the third year for the population generated through the first selection cycle (during the first and second years).
- Phenotype in location 3 only, in the thrd year, the expensive way.

```{r}
phenotype(plotType="Late", locations=3, years=3)
```

- We use pedigree information for relationship matrix, to estimate the genotypic values in the current breeding population.
- Here, all phenotypic information we obtained during breeding is used in the model.

```{r}
predictValue(sharingInfo="pedigree")
select()
```

- Self-fertilize selected individuals to create 120 progeny

```{r}
selfFertilize(nProgeny=120)
```

Genomic selection

- Get marker genotype in all population used in the breeding scheme.

```{r}
genotype()
```

- We specify locations 3 to only use trials from that location for training.
- We predict the genotypic values in the latest breeding population based on genomic prediction (GBLUP).

```{r}
predictValue(sharingInfo="markers", locations=3)
select()
cross()
```

## Look over the simulation result

We can check the simulation result. The thick line shows the mean value of five simulation trials. The thin lines show results of the five simulation trials, respectively.

```{r}
plotData()
```

Note that we can compare multiple results of other breeding strategies if we specify as so.



